controller/controller-worker/src/main/java/com/pinterest/rocksplicator/controller/tasks/AddHostTask.java

```diff
@@ -0,0 +1,229 @@
+
+import com.pinterest.rocksdb_admin.thrift.Admin;
+import com.pinterest.rocksdb_admin.thrift.BackupDBRequest;
+import com.pinterest.rocksdb_admin.thrift.RestoreDBRequest;
+import com.pinterest.rocksplicator.controller.bean.ClusterBean;
+import com.pinterest.rocksplicator.controller.bean.HostBean;
+import com.pinterest.rocksplicator.controller.bean.Role;
+import com.pinterest.rocksplicator.controller.bean.SegmentBean;
+import com.pinterest.rocksplicator.controller.bean.ShardBean;
+import com.pinterest.rocksplicator.controller.config.ConfigParser;
+import com.pinterest.rocksplicator.controller.util.AdminClientFactory;
+import com.pinterest.rocksplicator.controller.util.ShardUtil;
+import com.pinterest.rocksplicator.controller.util.ZKUtil;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.thrift.TException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import javax.inject.Inject;
+
+/**
+ * This task adds a new host to the given cluster. It consists of the following steps:
+ *
+ * 1) Check if the host to add is up and running. Fail if not.
+ * 2) Assign shard which doesn't have enough replicas to the new host
+ * 3) Upload shard data to the new host
+ * 4) Update cluster config in zookeeper to reflect the change
+ *
+ * @author Ang Xu (angxu@pinterest.com)
+ */
+public class AddHostTask extends TaskBase<AddHostTask.Param> {
+  private static final Logger LOG = LoggerFactory.getLogger(AddHostTask.class);
+  private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyymmdd-hhmmss");
+
+
+  @Inject
+  private CuratorFramework zkClient;
+
+  @Inject
+  private AdminClientFactory clientFactory;
+
+  public AddHostTask(String hostIp,
+                     int hostPort,
+                     String hostAZ,
+                     String hdfsDir,
+                     int rateLimitMbs) {
+    this(new Param()
+        .setHostToAdd(
+            new HostBean()
+                .setIp(hostIp)
+                .setPort(hostPort)
+                .setAvailabilityZone(hostAZ)
+        )
+        .setHdfsDir(hdfsDir)
+        .setRateLimitMbs(rateLimitMbs)
+    );
+  }
+
+  public AddHostTask(Param param) {
+    super(param);
+  }
+
+  @Override
+  public void process(Context ctx) throws Exception {
+    final String clusterName = ctx.getCluster();
+    final String hdfsDir = getParameter().getHdfsDir();
+    final HostBean hostToAdd = getParameter().getHostToAdd();
+    final int rateLimitMbs = getParameter().getRateLimitMbs();
+
+    InetSocketAddress hostAddr = new InetSocketAddress(hostToAdd.getIp(), hostToAdd.getPort());
+    Admin.Client client = clientFactory.getClient(hostAddr);
```

Comment:
Absolutely. Caching can be handled within AdminClientFactory which hasn't been implemented yet.